### DISCUSSION TITLE
Prevent losing draft comments on navigation or if a window is closed

### DISCUSSION THREAD

**Discussion by breilly2** [community/community#68448] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

When adding a comment in the "Files changed" area of a PR, anything entered is lost if the window is accidentally closed or another page is visited (even switching between Conversation/Commits/Checks/Files changed tabs). There is no warning alert on window close, and the draft unsubmitted comment is not saved anywhere (even browser local storage) and/or is not easily discovered. This can lead to a frustrating loss of work and time.

I've possibly seen _some_ evidence of unsubmitted comment text on single-line comments being saved, but I don't know if it's GitHub or Chrome (or a Chrome extension) doing that. If it is GitHub, the draft comment is completely undiscoverable and seems to not work at all for comments being added to multiple lines of code in the PR.

Possible enhancements are:
* Present a warning on navigate/close if there are form elements that have been filled out but not saved.
* Save the comment text somewhere and show an indication of unsubmitted comments when viewing the PR. (For comparison, this seems to be the behavior for the comment associated with the review.)

**Comment by liach** [comment]
I think this problem happens when I type in the comment box too. Sometimes to write pseudocode, I need to open other tabs and accidentally close this one when I navigate back/view changed files; and GitHub just dumps all my efforts.

I strongly agree that either prompting "are you sure to leave and discard unsaved changes" or saving in a cache works well to resolve this issue.

**Comment by aaron-williamson-kr** [comment]
Just lost an hour writing a comment, clicked a hyperlink in said comment to ensure it went to the right place and it opened in the same window rather than a new window. Lost everything. Really frustrating

**Comment by vic-bart** [comment]
Spent >1 hr writing a comment for an issue and ended up purging 20+ tabs I had opened, so I didn't realise I accidently closed the drafted comment, which hadn't been sent. Would've helped enormously if a dialogue window popped up warning me that I was drafting a comment before closing the tab (see example).
![image](https://github.com/user-attachments/assets/b469e9fd-e62e-44e7-90e4-251fee98933c)

**Reply by Chealer** [reply]
So for those who landed here, this is a mere _discussion_, not a ticket. It appears that the one person who was smart enough to create an ITS for GitHub ended up as a Microsoft employee, so even [the unofficial tracker he had created](https://github.com/isaacs/github/issues/376) is now read-only and basically useless.

What @vic-bart suggests does show (although apparently not all the time). Moreover, a draft is actually saved. It seems to work well on Microsoft Edge, but with Firefox 142, there appears to be a limit to the length of the comment. If a comment consists of 53+ times the line:

> This line consists of 37 characters.

…Firefox will quietly lose it. 52 times and less is fine though.

In other words, with at least Firefox, you will lose your most valuable comments, even if it's just for a requested OS restart, and sometimes without warning.😓 This is easy to reproduce, using Firefox’s _Unload Tab_ function.

**Comment by kswamin131** [comment]
I HATE doing PRs on GitHub for this very reason. 
For context this is also an issue in the "Conversations" section, which is where I draft the steps for testing my PRs.
I've lost so many hours or days of work if the site refreshes or I accidentally close or in the latest case, GitHub had issues creating the PR due to a duplicate older PR name and deleted the ENTIRE NEW PR instead of giving a warning before navigating away. Can we fix this at all and cache this data somewhere? SourceTree/Bitbucket/JIRA doesn't have these issues... and that's saying something

**Comment by owocado** [comment]
This would be immensely helpful. I had typed up a long draft for github issue but my laptop battery reached 0% and it shut down ( I somehow missed it in midst of doing multi-tasking multiple things), when I started it back and restarted the browser, the issues page loaded but empty content.

**Reply by Chealer** [reply]
This discussion is about merge requests. For tickets, see #42581.

**Comment by shanike** [comment]
Happened to me just now.

I was writing an invested and detailed comment, and paused to check some email. By mistake closed the tab and now seem to have lost the comment...

As mentioned in comments above me, a browser alert would be very helpful.

**Comment by mstyura** [comment]
Oh, this just happen to me - not due to I mis clicked something, but rather stupid company policy forcibly restarted my Chrome just in the middle of typing.

**Comment by aaronwng** [comment]
Happened to me when accidentally closed the window

**Comment by kiankyars** [comment]
same thing happened to me

**Comment by DerGut** [comment]
I think similar limitations exist for any text input in Github which make it hard for me to keep my entire workflow in the Github UI. Instead, I tend to draft out more elaborate texts in a local editor, iterate over them, and paste them into Github's UI when they're ready for submission.

If I forget to start locally, or if my text grows longer than anticipated, there's always the risk of losing my progress. And this happens at least once a month for various reasons (way too often).

---

This also means that there's not an easy way to work on drafts from multiple devices before submission. While draft PRs exist, I'm not aware of draft issues or draft comments. Draft PRs are also immediately made public.

I would find such a feature very useful! Partially because I haven't found a good workaround for this yet. My cross-device text editor of choice doesn't support Markdown which makes it very annoying to use for this use case. I hate to say it, but Confluence makes this a lot easier where I can start creating a synced draft page from any device, and then publish it from another...

**Comment by wolfschaf** [comment]
Just wanna say, I hate everything, since I lost an hour too just by misclicking once right after finishing....

**Reply by rlockerl** [reply]
i was enraged when i lost a one-sentence comment by simply clicking on another tab within the same PR.  i can't imagine losing an hour's worth of work.  how has this not yet been addressed???

**Reply by Chealer** [reply]
@rlockerl: it gets much worse.
Ironically, GitHub seems to stay this way due to its highly commercial nature. If you need a quality platform, look for one with an ITS. GitLab has a basic one.

**Comment by simondeschenes** [comment]
It just happened to me. I was typing a very long detailed comment and had been typing it for about 2 hours doing research to make sure I got the details right at the same time and I lost absolutely everything.

I would have assumed that website like github would have been smart enough to keep a copy of my in-progress comments in a cookie that is saved every few minutes. But it appears it isn't the case.

I just lost everything and I am not ready to start typing my issue report all over again, I am too infuriated.

**Reply by Chealer** [reply]
My sympathies, but this discussion is about merge requests. For tickets, see https://github.com/orgs/community/discussions/42581.

**Comment by noahzgordon** [comment]
Hi all. Chiming in because I think this problem has actually gotten worse in recent weeks. I don't remember this happening earlier, but now simply closing the "Submit Review" panel by (often accidentally) clicking Escape, or clicking elsewhere on the page, completely blanks out the input. This is very user-unfriendly and unintuitive. Could you at least preserve the content by keeping the element in the DOM when it is hidden?

**Comment by MarcoCarandenteDCL** [comment]
+1 on fixing this, absolutely bonkers that it's been going for so long now
